src/main/java/org/sonarlint/intellij/trigger/SonarLintSubmitter.java

```diff
@@ -48,36 +52,61 @@ public OpenFilesSubmitter(Project project, SonarLintConsole console, FileEditorM
     this.utils = utils;
   }
 
-  public void submit(TriggerType trigger) {
+  public void submitOpenFilesAuto(TriggerType trigger) {
+    if (!globalSettings.isAutoTrigger()) {
+      return;
+    }
     VirtualFile[] openFiles = editorManager.getOpenFiles();
-    submitFiles(openFiles, trigger);
+    submitFiles(openFiles, trigger, true, false);
   }
 
-  public void submitIfAutoEnabled(TriggerType trigger) {
-    if (!globalSettings.isAutoTrigger()) {
-      return;
+  /**
+   * Submit files for analysis.
+   * @param files Files to be analyzed. If empty, nothing is done and an empty AnalysisResult is returned.
+   * @param trigger What triggered the analysis
+   * @param autoTrigger Whether the analysis was triggered automatically. It affects the filter of files that should be analysed.
+   * @param manual Whether this is a user-initiated action. If it is, it will start running in foreground and will consume the single slot, updating
+   *               the status of the associated actions / icons.
+   */
+  public CompletableFuture<AnalysisResult> submitFiles(VirtualFile[] files, TriggerType trigger, boolean autoTrigger, boolean manual) {
+    Multimap<Module, VirtualFile> filesByModule = filterAndgetByModule(files, autoTrigger);
+
+    if (!filesByModule.isEmpty()) {
+      console.debug("Trigger: " + trigger);
+
+      for (Module m : filesByModule.keySet()) {
+        if (manual) {
+          return sonarLintJobManager.submitManual(m, filesByModule.get(m), trigger);
+        } else {
+          return sonarLintJobManager.submitBackground(m, filesByModule.get(m), trigger);
+        }
+      }
     }
-    submit(trigger);
+
+    CompletableFuture<AnalysisResult> future = new CompletableFuture<>();
+    future.complete(new AnalysisResult(0, Collections.emptyMap()));
+    return future;
   }
 
-  private void submitFiles(VirtualFile[] files, TriggerType trigger) {
+  private Multimap<Module, VirtualFile> filterAndgetByModule(VirtualFile[] files, boolean autoTrigger) {
     Multimap<Module, VirtualFile> filesByModule = HashMultimap.create();
 
     for (VirtualFile file : files) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the total number of break and continue statements in this loop to use at most one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS135)